server/controllers/stock/reports/stock/exit_report.js

```diff
@@ -0,0 +1,166 @@
+const {
+  _, db, ReportManager, pdfOptions, STOCK_EXIT_REPORT_TEMPLATE,
+} = require('../common');
+
+const StockExitToPatient = require('./exit/exitToPatient');
+const StockExitToService = require('./exit/exitToService');
+const StockExitToDepot = require('./exit/exitToDepot');
+const StockExitToLoss = require('./exit/exitToLoss');
+
+/**
+ * @method stockExitReport
+ *
+ * @description
+ * This method builds the stock exit report as either a JSON, PDF, or HTML
+ * file to be sent to the client.
+ *
+ * GET /reports/stock/exit
+ */
+function stockExitReport(req, res, next) {
+  let report;
+
+  let data = {};
+  let params = {};
+  const options = req.query;
+  const optionReport = _.extend(options, pdfOptions, {
+    filename : 'REPORT.STOCK.EXIT_REPORT',
+  });
+
+  // set up the report with report manager
+  try {
+    report = new ReportManager(STOCK_EXIT_REPORT_TEMPLATE, req.session, optionReport);
+    params = JSON.parse(req.query.params);
+  } catch (e) {
+    return next(e);
+  }
+
+  /**
+   * get into the data object :
+   * - exitToPatient : stock exit to patient data
+   * - exitToService : stock exit to service data
+   * - exitToDepot : stock exit to depot data
+   * - exitToLoss : stock exit as loss
+   */
+  try {
+    data = collect(params);
+  } catch (e) {
+    return next(e);
+  }
+
+  return fetchDepotDetails(params.depotUuid)
+    .then(depot => {
+      params.depotName = depot.text;
+      return data;
+    })
+    .then(groupCollection)
+    .then((bundle) => {
+      _.extend(bundle, params);
+
+      return report.render(bundle);
+    })
+    .then((result) => {
+      res.set(result.headers).send(result.report);
+    })
+    .catch(next)
+    .done();
+}
+
+/**
+ * fetchDepotDetails
+ * @param {number} depotUuid depot uuid
+ */
+function fetchDepotDetails(depotUuid) {
+  const query = 'SELECT text FROM depot WHERE uuid = ?';
+  return db.one(query, [db.bid(depotUuid)]);
+}
+
+/**
+ * @function groupCollection
+ * @description group collected data by inventory
+ */
+function groupCollection(exitCollection) {
+  const collection = {};
+
+  // exit to patient
+  collection.exitToPatient = _.groupBy(exitCollection.exitToPatient, row => row.text);
+  collection.exitToPatient = _.map(collection.exitToPatient, formatExit);
+  collection.exitToPatientEmpty = (Object.keys(collection.exitToPatient).length === 0);
+
+  // exit to service
+  collection.exitToService = _.groupBy(exitCollection.exitToService, row => row.text);
```

Comment:
You could avoid repeating code by sacrificing your `collection` object structure.  For example, if you used nested objects, you could do this:
```js
function formatAndCombine(data) {
  return _.chain(data)
    .groupBy('text')
    .map(formatExit)
    .value();
}

collection.toServices = formatAndCombine(exitCollection.exitToService);
collection.toServicesEmpty = _.size(collection.toServices) === 0;
```

You could even reduce it more:
```js
function formatAndCombine(data) {
  const aggregate = _.chain(data)
    .groupBy('text')
    .map(formatExit)
    .value();

  return { data : aggregate, isEmpty : _.size(aggregate) === 0 };
}

// note - this is now an object with properties .data and .isEmpty
collection.toServices = formatAndCombine(exitCollection.exitToService);
collection.toPatients = formatAndCombine(exitCollection.exitToPatient);
```

Just a suggestion!